lib/liquid/swap/ReverseSwapTree.ts

```diff
@@ -0,0 +1,118 @@
+import ops from '@boltz/bitcoin-ops';
+import { crypto, script } from 'bitcoinjs-lib';
+import { reverseBuffer } from 'liquidjs-lib/src/bufferutils';
+import { getHexBuffer } from '../../Utils';
+import { Tapleaf } from '../../consts/Types';
+import bitcoinReverseSwapTree from '../../swap/ReverseSwapTree';
+import { leafVersionLiquid } from '../../swap/TaprootUtils';
+import { assignTreeProbabilities, sortTree } from '../TreeSort';
+import { getScriptIntrospectionValues } from '../Utils';
+import liquidOps from '../consts/Ops';
+import { LiquidSwapTree } from '../consts/Types';
+
+enum Feature {
+  ClaimCovenant,
+}
+
+type ClaimCovenant = {
+  assetHash: string;
+  outputScript: Buffer;
+  expectedAmount: number;
+};
+
+type FeatureOption = { type: Feature } & ClaimCovenant;
+
+const claimCovenantOutputIndex = script.number.encode(0);
+
+const createClaimCovenantLeaf = (
+  preimageHash: Buffer,
+  assetHash: string,
+  outputScript: Buffer,
+  expectedAmount: number,
+): Tapleaf => {
+  const userOutput = getScriptIntrospectionValues(outputScript);
+
+  const amountBuffer = Buffer.alloc(8);
+  amountBuffer.writeBigUint64LE(BigInt(expectedAmount));
+
+  return {
+    version: leafVersionLiquid,
+    output: script.compile([
+      ops.OP_SIZE,
+      script.number.encode(32),
+      ops.OP_EQUALVERIFY,
+      ops.OP_HASH160,
+      crypto.ripemd160(preimageHash),
+      ops.OP_EQUALVERIFY,
+
+      claimCovenantOutputIndex,
+      liquidOps.OP_INSPECTOUTPUTSCRIPTPUBKEY,
+      script.number.encode(userOutput.version),
+      ops.OP_EQUALVERIFY,
+      userOutput.script,
+      ops.OP_EQUALVERIFY,
+
+      claimCovenantOutputIndex,
+      liquidOps.OP_INSPECTOUTPUTASSET,
+      ops.OP_DROP,
```

Comment:
Is it definitely safe to just compare the asset hash returned by `INSPECTOUTPUTASSET` without checking the asset prefix, too ?

It might be safe to compare value returned by INSPECTOUTPUTVALUE without checking the commitment prefix, because unconfidential value will be size 8 and confidential will be size 32

But the asset size is 32 bytes regardless if confidential or not, and as I remember in the blinding process there's 'balancing' values are added to the blinders of the last output

My understanding is that it is not required for these 'balancing' values to be added to any particular output, and maybe that balancing values can be also somehow distributed between different outputs ?

Might it be that it is possible to craft an arbitrary value for the (blinded) asset in one of the outputs ? Like, the covenant checks for unblinded asset value, ignores the commitment prefix, but the blinded value with the same bytes as for unblinded asset is crafted, while it is actually a blinded value?

I cannot say that I really understand the math behind the blinding process, so I wanted to ask someone knowledgeable in this, tagging them.

@apoelstra @roconnor-blockstream

Reply:
Thanks for the explanation !

Reply:
Isn't it "finding privkey for arbitrary pubkey" rather than sha2 collision ? If I understand correctly, you would need to get the binary representation of `SWU(y) + rG` to be equal to `x`, `SWU(y)` can be fixed, and that means you need to find `r` such as `rG = x - SWU(y)` (assuming that `y` can be interpreted as a point on the curve).

Reply:
Ah, yep, that would also do it.

Reply:
Thank you guys for the independent review, glad to hear what we are doing is safe :pray:

Reply:
> In short, yes, it is safe. But the argument for safety is complicated and crypto-heavy so if it's possible to avoid relying on it I would.

As @tiero said, checking explicitly for the prefix fixes the collision concern. That's what we'll do